flecsi/data/hpx/policy.hh

```diff
@@ -0,0 +1,396 @@
+    {
+      auto comm = flecsi::run::context::instance().world_comm();
+      auto f = util::hpx::all_to_allv(
+        [&](int r, int) -> auto & {
+          static const std::vector<std::size_t> empty;
+          const auto i = grouped_shared_entities.find(r);
+          return i == grouped_shared_entities.end() ? empty : i->second;
+        },
+        comm);
+
+      std::size_t r = 0;
+      for(auto & v : f.get()) {
+        if(!v.empty())
+          remote_ghost_entities.try_emplace(r, std::move(v));
+        ++r;
+      }
+    }
+
+    // We need to figure out the max local source index in order to give correct
+    // nelems when calling region::get_storage().
+    for(const auto & [rank, indices] : remote_ghost_entities) {
+      max_local_source_idx = (std::max)(max_local_source_idx,
+        *std::max_element(indices.begin(), indices.end()));
+    }
+    max_local_source_idx += 1;
+
+    // We need to reserve enough memory for MPI requests used in operator().
+    // This only need to be calculated once.
+    auto nrecvs = std::transform_reduce(destination.ghost_ranges.begin(),
+      destination.ghost_ranges.end(),
+      0,
+      std::plus<>(), // this uses C++14 std::plus<void> where T is deduced.
+      [](const auto & p) { return p.second - p.first; });
+    auto nsends = std::transform_reduce(remote_ghost_entities.begin(),
+      remote_ghost_entities.end(),
+      0,
+      std::plus<>(), // this uses C++14 std::plus<void> where T is deduced.
+      [](const auto & p) { return p.second.size(); });
+
+    nreqs = nrecvs + nsends;
+  }
+
+  // called with each field (and field_id_t) on the entity, for example, one
+  // for pressure, temperature, density etc.
+  void operator()(field_id_t data_fid) const {
+
+    auto source_storage =
+      source.r->get_storage<std::byte>(data_fid, max_local_source_idx);
+    auto destination_storage = destination.get_storage<std::byte>(data_fid);
+
+    // FIXME: should we assert(source_type_size == dest_type_size)?
+    auto type_size = source.r->get_field_info(data_fid)->type_size;
+
+    auto remote_sources =
+      destination.get_storage<const points::Value>(meta_fid);
+
+    auto comm = flecsi::run::context::instance().world_channel_comm();
+
+    using data_type = ::hpx::serialization::serialize_buffer<std::byte>;
+    std::vector<::hpx::future<void>> requests;
+    requests.reserve(nreqs);
+
+    for(const auto & [begin, end] : destination.ghost_ranges) {
+      for(auto ghost_idx = begin; ghost_idx < end; ++ghost_idx) {
+        auto source_rank = remote_sources[ghost_idx].first;
+        requests.push_back(
+          ::hpx::collectives::get<data_type>(
+            comm, ::hpx::collectives::that_site_arg(source_rank))
+            .then([&](auto && f) {
+              auto && data = f.get();
+              std::copy(data.data(),
+                data.data() + data.size(),
+                destination_storage.data() + ghost_idx * type_size);
+            }));
```

Comment:
How plausible do you think it might be to factor out differences like this between the MPI and HPX backends so that, say, a template could be instantiated with either behavior?  Since we don't claim any support for mixing backends, we _could_ just use `#if` to select the implementation within each function, but that's fairly hideous.

Reply:
Same as above. This needs separate investigation.